enhancements/baremetal/baremetal-keepalived-unicast.md

```diff
@@ -0,0 +1,191 @@
+Keepalived uses the `virtual_router_id` field to specify to which VRRP router the [instance](https://www.keepalived.org/doc/configuration_synopsis.html#vrrp-instance-definitions-synopsis) belongs,`virtual_router_id` range is 1 - 255.
+Each baremetal deployment includes two virtual routers:
+* API
+* Ingress
+
+Although the `virtual_router_id` values for both API and Ingress are calculated using [checksum function based on cluster name](https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/config/node.go#L162-#L171) we can still end up in a situation where virtual routers from different deployments or services share the same `virtual_router_id`.
+
+### Implementation Details/Notes/Constraints
+
+Unlike multicast mode, in unicast we should specify and keep up to date all the group members in 'unicast_peer' section at [Keepalived conf](https://www.keepalived.org/manpage.html) for each VIP. Lets first present briefly the supported VIPs and the group members for each one of them.
+
+As mentioned above, `api-vip` and `ingress-vip` are supported to enable high availability.
+
+The `api-vip` should resides on the bootstrap instance during the bootstrap phase and move to one of the control plane nodes as soon as one of the control plane nodes can host it.
+The group members for the `api-vip` should include the bootstrap node and the control plane nodes.
+
+As per the `ingress-vip`, it should resides on a node (either control or compute) that runs a router pod, so for that case the group members should include all control and compute nodes.
+
+#### api-vip implementation
+
+Since the bootstrap IP address is not known until the server is up, a method for 
+retrieving that node will need to be implemented to allow the control plane nodes to include the 
+bootstrap IP address in their Keepalived configuration. 
+
+##### Bootstrap node
+
+The bootstrap [Keepalived](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml) instance runs as static pod and the relevant [Keepalived configuration](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.conf.tmpl) rendered by the [baremetal-runtimecfg](https://github.com/openshift/baremetal-runtimecfg).
+
+The bootstrap Keepalived pod includes:
+- [Keepalived container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L35-#L93)
+- [TCP server container](https://github.com/openshift/machine-config-operator/blob/4c04978faf8ce36af046e18c7722d03f6e60133e/manifests/baremetal/keepalived.yaml#L27#L34) that will be responsible to return the bootstrap IP address.
```

Comment:
This is probably outdated?

Reply:
When running keepalived (for API-VIP) in unicast mode during bootstrap phase we need to set the following : 

*  In bootstrap node, add masters nodes IPs to 'unicast_peer' in keepalived conf 
*  In master nodes, add other masters IPs + bootstrap node IP address to  'unicast_peer' in keepalived conf.

We retrieve the master nodes details from OCP-API, and we need to find a way to retrieve somehow the bootstrap IP address, so, we decided to run a TCP server only on the bootstrap node for that purpose.

Do you know if the bootstrap IP address could be retrieved in some other way?

Reply:
The bootstrap node itself could provide it when generating manifests?

Reply:
@celebdor  Do you think ^^ is feasible?

Reply:
I don't know. Did you check the MCO bootstrap and the installer bootstrap to check it?

Reply:
I checked with MCO and installer teams, seems that the bootstrap IP could be retrieved without running a dedicated server in the bootstrap node for that purpose. 
One option is to read it from etcd-endpoints ConfigMap, I updated the enhancement (the code should be updated).

Reply:
Done